Google.PowerShell/Storage/GcsObject.cs

```diff
@@ -49,23 +49,53 @@ protected bool TestObjectExists(StorageService service, string bucket, string ob
         protected Object UploadGcsObject(
             StorageService service, string bucket, string objectName,
             Stream contentStream, string contentType,
-            PredefinedAclEnum? predefinedAcl, Dictionary<string, string> metadata)
+            PredefinedAclEnum? predefinedAcl, Dictionary<string, string> metadata,
+            string cacheControl = null, string contentDisposition = null,
+            string contentEncoding = null, string contentLanguage = null)
         {
-            // Work around an API wart. It is possible to specify content type via the API and also by
-            // metadata.
-            if (metadata != null && metadata.ContainsKey("Content-Type"))
-            {
-                metadata["Content-Type"] = contentType;
-            }
-
             Object newGcsObject = new Object
             {
                 Bucket = bucket,
                 Name = objectName,
                 ContentType = contentType,
-                Metadata = metadata
+                ContentEncoding = contentEncoding,
+                ContentDisposition = contentDisposition,
+                CacheControl = cacheControl,
+                ContentLanguage = contentLanguage
             };
 
+            if (metadata != null)
+            {
+                // Handles fixed-key metadata. Removes them so there won't be duplicate. See:
+                // https://cloud.google.com/storage/docs/metadata#mutable
+                if (metadata.ContainsKey("Content-Type"))
```

Comment:
This should probably be a list of constants and a loop (or the values of a Dictionary).

Alternatively, it could be moved inside of `GetFixedTypeMetadata`